File: src/cogs/Roulette.py
import discord
from discord.ext import commands
import random
from discord.ui import Button, View

from src.database.balance import update_balance, get_balance
from src.database.achievement import increment_stat, check_and_unlock_achievements, format_achievements_unlocks
from src.database.settings import get_language
from src.utils.i18n import t


class RouletteGameView(View):
    def __init__(self, players, entry_fee, lang):
        super().__init__(timeout=300)
        self.players = players
        self.entry_fee = entry_fee
        self.lang = lang
        self.alive_players = players.copy()
        self.cylinder = [False] * 6
        self.cylinder[random.randint(0, 5)] = True
        self.turn_index = 0
        self.pot = len(players) * entry_fee

    def get_current_player(self):
        return self.alive_players[self.turn_index % len(self.alive_players)]

    @discord.ui.button(label="trigger", style=discord.ButtonStyle.danger)
    async def trigger(self, interaction: discord.Interaction, button: Button):
        shooter = self.get_current_player()
        if interaction.user != shooter:
            return await interaction.response.send_message(t("roulette.not_your_turn", self.lang), ephemeral=True)
        bullet = self.cylinder.pop(0)
        if bullet:
            await interaction.response.send_message(t("roulette.bang_msg", self.lang, user=shooter.mention))
            self.stop()
            increment_stat(shooter.id, "roulette_lost")
            increment_stat(shooter.id, "roulette_spent", self.entry_fee)
            increment_stat(shooter.id, "roulette_money_lost", self.entry_fee)
            
            survivors = [p for p in self.alive_players if p != shooter]
            share = self.pot // len(survivors)
            text = t("roulette.survivors_win", self.lang, user=shooter.display_name, fee=self.entry_fee, count=len(survivors), share=share)
            embed = discord.Embed(title=t("roulette.finish_title", self.lang), description=text, color=discord.Color.red())
            await interaction.channel.send(embed=embed)
            for s in survivors:
                update_balance(s.id, share)
                increment_stat(s.id, "roulette_won")
                increment_stat(s.id, "roulette_spent", self.entry_fee)
                
                net_win = share - self.entry_fee
                if net_win > 0:
                    increment_stat(s.id, "roulette_money_won", net_win)
                    
                unlocks = check_and_unlock_achievements(s.id)
                if unlocks:
                    await interaction.channel.send(content=s.mention, embed=format_achievements_unlocks(unlocks, self.lang))
            return
        else:
            await interaction.response.send_message(t("roulette.clic_msg", self.lang, user=shooter.display_name))
            self.turn_index += 1
            next_player = self.get_current_player()

            embed = discord.Embed(description=t("roulette.next_turn", self.lang, user=next_player.mention),
                                  color=discord.Color.dark_grey())
            return await interaction.channel.send(embed=embed)


class JoinView(View):
    def __init__(self, ctx, amount, lang):
        super().__init__(timeout=60)
        self.ctx = ctx
        self.amount = amount
        self.lang = lang
        self.players = []

    @discord.ui.button(label="join", style=discord.ButtonStyle.blurple)
    async def join(self, interaction, button):
        if interaction.user in self.players:
            return await interaction.response.send_message(t("roulette.already_joined", self.lang), ephemeral=True)

        if get_balance(interaction.user.id) < self.amount:
            return await interaction.response.send_message(t("roulette.no_money", self.lang), ephemeral=True)

        update_balance(interaction.user.id, -self.amount)
        self.players.append(interaction.user)
        return await interaction.response.send_message(t("roulette.joined_msg", self.lang, user=interaction.user.display_name),
                                                ephemeral=False)

    @discord.ui.button(label="start", style=discord.ButtonStyle.green)
    async def start(self, interaction, button):
        if interaction.user != self.ctx.author:
            return await interaction.response.send_message(t("roulette.only_leader", self.lang),
                                                           ephemeral=True)
        if len(self.players) < 2:
            return await interaction.response.send_message(t("roulette.min_players", self.lang), ephemeral=True)
        self.stop()
        game_view = RouletteGameView(self.players, self.amount, self.lang)
        # Update button label
        for item in game_view.children:
            if isinstance(item, Button) and item.label == "trigger":
                item.label = t("roulette.trigger_label", self.lang)

        first_player = game_view.get_current_player()
        return await interaction.channel.send(
            t("roulette.start_announcement", self.lang, user=first_player.mention),
            view=game_view
        )


class Roulette(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command(name='roulette', aliases=['rr'])
    async def roulette(self, ctx, amount: int):
        """Roulette Russe. 1 chance sur 6 de mourir (et perdre la mise)."""
        lang = get_language(ctx.guild.id if ctx.guild else None)
        if amount <= 0:
            return await ctx.send(t("blackjack.invalid_bet", lang))
        
        view = JoinView(ctx, amount, lang)
        # Update button labels
        for item in view.children:
            if isinstance(item, Button):
                if item.label == "join": item.label = t("roulette.join_label", lang)
                if item.label == "start": item.label = t("roulette.start_label", lang)

        return await ctx.send(
            t("roulette.open_msg", lang, amount=amount, user=ctx.author.mention),
            view=view)


async def setup(bot):
    await bot.add_cog(Roulette(bot))

File: internal/cogs/roulette/roulette.go
package roulette

import (
	"strconv"
	"strings"

	"github.com/bwmarrin/discordgo"

	"guacagamblebot/internal/achievement"
	"guacagamblebot/internal/components"
	"guacagamblebot/internal/config"
	"guacagamblebot/internal/i18n"
	"guacagamblebot/internal/interaction"
	rlt "guacagamblebot/internal/service/roulette"
	"guacagamblebot/internal/store"
)

type Cog struct {
	store *store.Store
	cfg   *config.Config
	games map[int64]*rlt.Game // serverID -> game
}

func Register(r *interaction.Router, s *store.Store, cfg *config.Config) {
	c := &Cog{store: s, cfg: cfg, games: map[int64]*rlt.Game{}}
	r.Slash("roulette", "Roulette russe : rejoignez ou créez une partie.", c.onSlashMenu)
	r.Prefix("roulette", c.onPrefixMenu)
	r.Component("roulette", "new", c.onNewOpen)
	r.Component("roulette", "join", c.onJoin)
	r.Component("roulette", "start", c.onStart)
	r.Component("roulette", "trigger", c.onTrigger)
	r.Modal("roulette", "new_submit", c.onNewSubmit)
}

func (c *Cog) onSlashMenu(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	embed, comps := c.menu(lang, interaction.ToInt64(i.GuildID))
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseChannelMessageWithSource, embed, comps))
}

func (c *Cog) onPrefixMenu(b *interaction.Bot, s *discordgo.Session, m *discordgo.Message) {
	lang := c.store.GetLanguage(interaction.ToInt64(m.GuildID))
	embed, comps := c.menu(lang, interaction.ToInt64(m.GuildID))
	_, _ = s.ChannelMessageSendComplex(m.ChannelID, &discordgo.MessageSend{
		Embeds:     []*discordgo.MessageEmbed{embed},
		Components: comps,
	})
}

func (c *Cog) menu(lang string, serverID int64) (*discordgo.MessageEmbed, []discordgo.MessageComponent) {
	game := c.games[serverID]
	desc := i18n.T("roulette.open_msg", lang, map[string]any{"amount": 0, "user": "?"})
	if game != nil {
		playerMentions := ""
		for _, p := range game.Players {
			playerMentions += interaction.Mention(p.UserID) + " "
		}
		desc = i18n.T("roulette.open_msg", lang, map[string]any{
			"amount": game.EntryFee,
			"user":   interaction.Mention(game.LeaderID),
		})
		if len(game.Players) > 0 {
			desc += "\n\n**Players:** " + playerMentions
		}
	}

	embed := components.Embed(
		i18n.T("roulette.finish_title", lang),
		desc,
		0xe74c3c,
	)

	comps := []discordgo.MessageComponent{
		components.ActionRow(
			components.Button(i18n.T("roulette.join_label", lang), components.Encode("roulette", "new"), discordgo.PrimaryButton),
		),
	}
	return embed, comps
}

func (c *Cog) onNewOpen(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	modal := components.ModalResponse(
		components.Encode("roulette", "new_submit"),
		i18n.T("roulette.open_msg", lang, map[string]any{"amount": "", "user": ""}),
		components.TextInput("entry_fee", i18n.T("economy.quantity", lang), true, "100", discordgo.TextInputShort, 1, 12),
	)
	_ = b.Session.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: modal,
	})
}

func (c *Cog) onNewSubmit(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	serverID := interaction.ToInt64(i.GuildID)
	leaderID := interaction.ToInt64(interaction.UserID(i))
	values := interaction.ModalValues(i)

	entryFee, err := strconv.Atoi(strings.TrimSpace(values["entry_fee"]))
	if err != nil || entryFee <= 0 {
		interaction.RespondError(b, i, lang, "blackjack.invalid_bet")
		return
	}

	bal, err := c.store.GetBalance(leaderID)
	if err != nil || bal < entryFee {
		interaction.RespondError(b, i, lang, "roulette.no_money")
		return
	}

	if _, err := c.store.UpdateBalance(leaderID, -entryFee); err != nil {
		interaction.RespondError(b, i, lang, "roulette.no_money")
		return
	}

	game := rlt.NewGame(leaderID, entryFee)
	c.games[serverID] = game

	embed := components.Embed(
		i18n.T("roulette.finish_title", lang),
		i18n.T("roulette.open_msg", lang, map[string]any{
			"amount": entryFee,
			"user":   interaction.Mention(leaderID),
		})+"\n\n👤 "+interaction.Mention(leaderID),
		0xe74c3c,
	)
	comps := []discordgo.MessageComponent{
		components.ActionRow(
			components.Button(i18n.T("roulette.join_label", lang), components.Encode("roulette", "join"), discordgo.PrimaryButton),
			components.Button(i18n.T("roulette.start_label", lang), components.Encode("roulette", "start"), discordgo.SuccessButton),
		),
	}
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseChannelMessageWithSource, embed, comps))
}

func (c *Cog) onJoin(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	serverID := interaction.ToInt64(i.GuildID)
	userID := interaction.ToInt64(interaction.UserID(i))

	game, ok := c.games[serverID]
	if !ok || game == nil {
		interaction.RespondError(b, i, lang, "roulette.no_money")
		return
	}
	if game.IsActive {
		interaction.RespondError(b, i, lang, "roulette.already_joined")
		return
	}

	bal, err := c.store.GetBalance(userID)
	if err != nil || bal < game.EntryFee {
		interaction.RespondError(b, i, lang, "roulette.no_money")
		return
	}

	if err := game.AddPlayer(userID); err != nil {
		interaction.RespondError(b, i, lang, "roulette.already_joined")
		return
	}

	if _, err := c.store.UpdateBalance(userID, -game.EntryFee); err != nil {
		interaction.RespondError(b, i, lang, "roulette.no_money")
		return
	}

	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseChannelMessageWithSource,
			components.Embed(
				"",
				i18n.T("roulette.joined_msg", lang, map[string]any{"user": interaction.Mention(userID)}),
				0x2ecc71,
			),
			nil,
		))
}

func (c *Cog) onStart(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	serverID := interaction.ToInt64(i.GuildID)
	userID := interaction.ToInt64(interaction.UserID(i))

	game, ok := c.games[serverID]
	if !ok || game == nil {
		interaction.RespondError(b, i, lang, "roulette.no_money")
		return
	}

	if err := game.Start(userID); err != nil {
		switch err {
		case rlt.ErrNotLeader:
			interaction.RespondError(b, i, lang, "roulette.only_leader")
		case rlt.ErrMinPlayers:
			interaction.RespondError(b, i, lang, "roulette.min_players")
		default:
			interaction.RespondError(b, i, lang, "roulette.no_money")
		}
		return
	}

	cp := game.CurrentPlayer()
	embed := components.Embed(
		i18n.T("roulette.finish_title", lang),
		i18n.T("roulette.start_announcement", lang, map[string]any{"user": interaction.Mention(cp.UserID)}),
		0xe74c3c,
	)
	comps := []discordgo.MessageComponent{
		components.ActionRow(
			components.Button(i18n.T("roulette.trigger_label", lang), components.Encode("roulette", "trigger"), discordgo.DangerButton),
		),
	}
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}

func (c *Cog) onTrigger(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	serverID := interaction.ToInt64(i.GuildID)
	userID := interaction.ToInt64(interaction.UserID(i))

	game, ok := c.games[serverID]
	if !ok || game == nil || !game.IsActive {
		interaction.RespondError(b, i, lang, "roulette.not_your_turn")
		return
	}

	ok, result, survivors, share := game.Trigger(userID)
	if !ok {
		interaction.RespondError(b, i, lang, "roulette.not_your_turn")
		return
	}

	if result == "dead" {
		_ = achievement.IncrementStat(b.DB, userID, "roulette_lost", 1)
		_ = achievement.IncrementStat(b.DB, userID, "roulette_spent", game.EntryFee)
		_ = achievement.IncrementStat(b.DB, userID, "roulette_money_lost", game.EntryFee)

		desc := i18n.T("roulette.bang_msg", lang, map[string]any{"user": interaction.Mention(userID)})
		desc += "\n" + i18n.T("roulette.survivors_win", lang, map[string]any{
			"user":  interaction.Mention(userID),
			"fee":   game.EntryFee,
			"count": len(survivors),
			"share": share,
		})

		for _, s := range survivors {
			if _, err := c.store.UpdateBalance(s.UserID, share); err != nil {
				continue
			}
			_ = achievement.IncrementStat(b.DB, s.UserID, "roulette_won", 1)
			_ = achievement.IncrementStat(b.DB, s.UserID, "roulette_spent", game.EntryFee)
			net := share - game.EntryFee
			if net > 0 {
				_ = achievement.IncrementStat(b.DB, s.UserID, "roulette_money_won", net)
			}
		}

		embed := components.Embed(i18n.T("roulette.finish_title", lang), desc, 0xe74c3c)
		_ = b.Session.InteractionRespond(i.Interaction,
			components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, nil))

		for _, s := range survivors {
			unlocks, _ := achievement.CheckAndUnlock(b.DB, s.UserID)
			if len(unlocks) > 0 {
				interaction.SendAchievements(b, i, lang, unlocks)
			}
		}

		delete(c.games, serverID)
		return
	}

	cp := game.CurrentPlayer()
	desc := i18n.T("roulette.clic_msg", lang, map[string]any{"user": interaction.Mention(userID)})
	desc += "\n" + i18n.T("roulette.next_turn", lang, map[string]any{"user": interaction.Mention(cp.UserID)})

	embed := components.Embed(i18n.T("roulette.finish_title", lang), desc, 0x95a5a6)
	comps := []discordgo.MessageComponent{
		components.ActionRow(
			components.Button(i18n.T("roulette.trigger_label", lang), components.Encode("roulette", "trigger"), discordgo.DangerButton),
		),
	}
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}
